_drafts/arm-cortex-m-exceptions-and-nvic.md

```diff
@@ -0,0 +1,699 @@
+  volatile uint32_t *nvic_iser = (void *)0xE000E100;
+  *nvic_iser |= (0x1 << 9) | (0x1 << 10) | (0x1 << 11);
+
+  // Pend an interrupt
+  volatile uint32_t *nvic_ispr = (void *)0xE000E200;
+  *nvic_ispr |= (0x1 << 9) | (0x1 << 10) | (0x1 << 11);
+
+  // flush pipeline to ensure exception takes effect before we
+  // return from this routine
+  __asm("isb");
+}
+```
+
+Let's call `trigger_nvic_int9_int10_int11` and try it out!
+
+```
+(gdb) c
+Continuing.
+
+Program received signal SIGTRAP, Trace/breakpoint trap.
+TIMER2_IRQHandler () at ../../../main.c:81
+81    __asm("bkpt 5");
+(gdb)
+```
+
+So the external interrupt 10 (exception number 26) fired first. It has the same priority as NVIC Interrupt 11 but the ARM
+core prioritizes higher Exception Numbers first which is why 10 is the first one that runs. We
+would expect NVIC Interrupt 11 to run next.
+
+Let's check and see what info is in the `ICSR` register this time:
+
+```
+(gdb) p/x *(uint32_t*)0xE000ED04
+$9 = 0x41b81a
+(gdb) p/d (*(uint32_t*)0xE000ED04)&0xff
+$10 = 26
+(gdb) p/d (*(uint32_t*)0xE000ED04)>>12&0x1
+$11 = 1
+(gdb) p/d (*(uint32_t*)0xE000ED04)>>12&0xff
+$12 = 27
+```
+
+`VECTACTIVE` is 26 which matches what we expect. This time `VECTPENDING` is set too! The value is
+27 which confirms that External Interrupt 11 (27-16) should be the next one to fire.
+
+We can see all the NVIC interrupts that are pended by looking at the NVIC_ISPR. We should see bits
+9 and 11 set since those interrupts haven't run yet
+
+```
+(gdb) p/x *(uint32_t[16] *)0xE000E200
+$13 = {0xa00, 0x0 <repeats 15 times>}
+```
+
+Let's step through the rest of the code and confirm we see `bkpt 6` followed by `bkpt 4`:
+
+```
+(gdb) c
+Continuing.
+
+Program received signal SIGTRAP, Trace/breakpoint trap.
+RTC0_IRQHandler () at ../../../main.c:86
+86    __asm("bkpt 6");
+(gdb) next
+main (a=<optimized out>, argv=<optimized out>) at ../../../main.c:127
+127       bsp_board_led_invert(i);
+(gdb) c
+Continuing.
+
+Program received signal SIGTRAP, Trace/breakpoint trap.
+TIMER1_IRQHandler () at ../../../main.c:76
+76    __asm("bkpt 4");
+```
+
+## Closing
+
+I hope this post gave you a useful overview of how the ARM Cortex-M Exception model works. There's
+a lot of different reference manuals and books about the topic but I've always found it hard to
+find a single place that aggregates the useful information.
+
+Are there any other topics related to interrupts you'd like us to delve into? (No pun intended :D) Do you leverage any of
+ARMs fancy exception configuration features in your products? Let us know in the discussion area below!
+
+See anything you'd like to change? Submit a pull request or open an issue at [Github](https://github.com/memfault/interrupt)
+
+## Additional Reading
+
+If you'd like to read even more here's some other dicussions about Cortex-M exceptions
+that I've found to be interesting:
+
+- [Cortex-M Exception Handling](https://www.embeddedrelated.com/showarticle/878.php)
```

Comment:
This currently shows up as a title, for some reason.

Reply:
![image](https://user-images.githubusercontent.com/4192460/64193150-0854b800-ce31-11e9-948b-e2906c9c217b.png)

Reply:
oops thought i pushed a fix for that, done